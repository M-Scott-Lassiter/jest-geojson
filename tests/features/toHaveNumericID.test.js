const { goodGeometry, badGeometry, nestedGeometryCollection } = require('../geometries/data')

const validIDs = [0, -200, 200, Infinity, -Infinity]
const invalidSearchIDs = [
    '1',
    'Test 123',
    'Random ID String',
    '',
    NaN,
    null,
    true,
    false,
    { someProp: 'I am not GeoJSON', id: 4 },
    {},
    goodGeometry.lineString,
    goodGeometry.multiLineString,
    goodGeometry.multiPoint,
    goodGeometry.multiPolygon,
    goodGeometry.point,
    goodGeometry.polygon,
    nestedGeometryCollection,
    badGeometry.lineString,
    badGeometry.multiLineString,
    badGeometry.multiPoint,
    badGeometry.multiPolygon,
    badGeometry.point,
    badGeometry.polygon,
    badGeometry.geometryCollection,
    badGeometry.unrecognized
]
const invalidInputValues = [
    ...invalidSearchIDs,
    undefined,
    '',
    'Random Feature',
    JSON.stringify({
        type: 'Feature',
        geometry: null,
        properties: null
    })
]

describe('Valid Use Cases', () => {
    describe('Expect to pass with good features and IDs', () => {
        const testFeature = {
            type: 'Feature',
            geometry: goodGeometry.polygon,
            properties: null,
            id: null
        }

        describe('No Specific Callout: expect(testFeature).toHaveNumericID()', () => {
            test.each(validIDs)('ID: %p', (input) => {
                testFeature.id = input
                expect(testFeature).toHaveNumericID()
            })
        })

        describe('Specific Callout', () => {
            test.each(validIDs)('expect(testFeature).toHaveNumericID(%p)', (input) => {
                testFeature.id = input
                expect(testFeature).toHaveNumericID(input)
            })
        })

        describe('Specific Callout in Single Array Element', () => {
            test.each(validIDs)('expect(testFeature).toHaveNumericID([%p])', (input) => {
                testFeature.id = input
                expect(testFeature).toHaveNumericID([input])
            })
        })

        describe('Specific Callout RegExp', () => {
            test.each(validIDs)('expect(testFeature).toHaveNumericID(/%p/)', (input) => {
                testFeature.id = input
                expect(testFeature).toHaveNumericID(new RegExp(input))
            })
        })

        describe('Specific Callout RegExp in Single Array Element', () => {
            test.each(validIDs)('expect(testFeature).toHaveNumericID([/%p/])', (input) => {
                testFeature.id = input
                expect(testFeature).toHaveNumericID([new RegExp(input)])
            })
        })
    })

    describe('Expect to pass Multiple Array Value Checking for Numeric ID', () => {
        const searchArray = [
            [[1, 719, 0, 123]],
            [[-5, /7/, 15]],
            [[/719/, /123/, /\bString\b/]],
            [[/72[0-9]/, /\bString\b/, /71[0-9]/, 719]]
        ]

        test.each(searchArray)('ID Search Array: %p', (input) => {
            const testFeature = {
                type: 'Feature',
                geometry: goodGeometry.polygon,
                properties: null,
                id: 719
            }
            expect(testFeature).toHaveNumericID(input)
        })

        test('Empty Array for Optional Argument. ID: %p', () => {
            const testFeature = {
                type: 'Feature',
                geometry: null,
                properties: null,
                id: 719
            }
            expect(testFeature).toHaveNumericID([])
        })
    })
})

describe('Invalid Use Cases', () => {
    describe('Expect to fail with bad inputs:', () => {
        test.each(invalidInputValues)('expect(%p).not.toHaveNumericID()', (badInput) => {
            expect(badInput).not.toHaveNumericID()
        })
    })

    test('Fails when provided invalid feature', () => {
        const testFeature = {
            type: 'Feature',
            properties: null
        }
        expect(testFeature).not.toHaveNumericID()
    })

    test('Missing ID, searchID is empty array.', () => {
        const testFeature = {
            type: 'Feature',
            geometry: null,
            properties: null
        }
        expect(testFeature).not.toHaveNumericID([])
    })

    describe('Known good Feature with ID That Does Not Match:', () => {
        test.each([0, -500, 2000])('ID: %p', (input) => {
            const testFeature = {
                type: 'Feature',
                geometry: goodGeometry.polygon,
                properties: null,
                id: input
            }
            const badID = input + 1

            expect(testFeature).not.toHaveNumericID(badID)
            expect(testFeature).not.toHaveNumericID([badID])
            expect(testFeature).not.toHaveNumericID(new RegExp(badID))
            expect(testFeature).not.toHaveNumericID([new RegExp(badID)])
        })
    })

    describe('Expect to fail with non-numeric ID:', () => {
        test.each(invalidSearchIDs)('expect(testFeature).not.toHaveNumericID(%p)', (badInput) => {
            const testFeature = {
                type: 'Feature',
                geometry: goodGeometry.polygon,
                properties: null,
                id: badInput
            }

            expect(testFeature).not.toHaveNumericID(badInput)
            expect(testFeature).not.toHaveNumericID([badInput])
            expect(testFeature).not.toHaveNumericID(new RegExp(badInput))
            expect(testFeature).not.toHaveNumericID([new RegExp(badInput)])
        })
    })

    test('Fails when no ID provided to valid feature', () => {
        const testFeature = {
            type: 'Feature',
            geometry: null,
            properties: null
        }
        expect(testFeature).not.toHaveNumericID()
    })
})

describe('Error Snapshot Testing. Throws error:', () => {
    test('Valid use case passes', () => {
        const testFeature = {
            type: 'Feature',
            geometry: goodGeometry.polygon,
            properties: null,
            id: 755
        }
        expect(() => expect(testFeature).not.toHaveNumericID()).toThrowErrorMatchingSnapshot()
    })

    test('Invalid input to matcher', () => {
        expect(() =>
            expect(goodGeometry.lineString).toHaveNumericID()
        ).toThrowErrorMatchingSnapshot()
    })

    test('Non-numeric search ID', () => {
        const testFeature = {
            type: 'Feature',
            geometry: goodGeometry.polygon,
            properties: null,
            id: 1
        }
        expect(() => expect(testFeature).toHaveNumericID('1')).toThrowErrorMatchingSnapshot()
    })

    test('No ID Is Present', () => {
        const testFeature = {
            type: 'Feature',
            geometry: null,
            properties: null
        }
        expect(() => expect(testFeature).toHaveNumericID()).toThrowErrorMatchingSnapshot()
        expect(() => expect(testFeature).toHaveNumericID([])).toThrowErrorMatchingSnapshot()
    })

    test('ID Present, Empty Array Search', () => {
        const testFeature = {
            type: 'Feature',
            geometry: null,
            properties: null,
            id: 22
        }
        expect(() => expect(testFeature).not.toHaveNumericID([])).toThrowErrorMatchingSnapshot()
    })

    test('ID Present, Search for Wrong ID', () => {
        const testFeature = {
            type: 'Feature',
            geometry: null,
            properties: null,
            id: 22
        }
        expect(() => expect(testFeature).toHaveNumericID(21)).toThrowErrorMatchingSnapshot()
        expect(() => expect(testFeature).toHaveNumericID([23])).toThrowErrorMatchingSnapshot()
    })

    test('String ID', () => {
        const testFeature = {
            type: 'Feature',
            geometry: null,
            properties: null,
            id: '22'
        }
        expect(() => expect(testFeature).toHaveNumericID('22')).toThrowErrorMatchingSnapshot()
    })
})
