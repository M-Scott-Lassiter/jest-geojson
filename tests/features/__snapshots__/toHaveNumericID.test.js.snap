// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Snapshot Testing. Throws error: ID Present, Empty Array Search 1`] = `
"[2mexpect([22m[31mFeatureObject[39m[2m).not.toHaveNumericID([22m[32mSearchID[39m[2m)[22m

Expected input to not be a valid GeoJSON Feature object with number type ID.

Received:  [31m{\\"geometry\\": null, \\"id\\": 22, \\"properties\\": null, \\"type\\": \\"Feature\\"}[39m"
`;

exports[`Error Snapshot Testing. Throws error: ID Present, Search for Wrong ID 1`] = `
"[2mexpect([22m[31mFeatureObject[39m[2m).toHaveNumericID([22m[32mSearchID[39m[2m)[22m

Did not find the numeric ID 21.

Received:  [31m{\\"geometry\\": null, \\"id\\": 22, \\"properties\\": null, \\"type\\": \\"Feature\\"}[39m"
`;

exports[`Error Snapshot Testing. Throws error: ID Present, Search for Wrong ID 2`] = `
"[2mexpect([22m[31mFeatureObject[39m[2m).toHaveNumericID([22m[32mSearchID[39m[2m)[22m

Did not find the numeric ID 23.

Received:  [31m{\\"geometry\\": null, \\"id\\": 22, \\"properties\\": null, \\"type\\": \\"Feature\\"}[39m"
`;

exports[`Error Snapshot Testing. Throws error: Invalid input to matcher 1`] = `
"[2mexpect([22m[31mFeatureObject[39m[2m).toHaveNumericID([22m[32mSearchID[39m[2m)[22m

Must have a type property with value 'Feature'

Received:  [31m{\\"coordinates\\": [[101, 0], [102, 1]], \\"type\\": \\"LineString\\"}[39m"
`;

exports[`Error Snapshot Testing. Throws error: No ID Is Present 1`] = `
"[2mexpect([22m[31mFeatureObject[39m[2m).toHaveNumericID([22m[32mSearchID[39m[2m)[22m

Feature does not have an \\"id\\" member.

Received:  [31m{\\"geometry\\": null, \\"properties\\": null, \\"type\\": \\"Feature\\"}[39m"
`;

exports[`Error Snapshot Testing. Throws error: No ID Is Present 2`] = `
"[2mexpect([22m[31mFeatureObject[39m[2m).toHaveNumericID([22m[32mSearchID[39m[2m)[22m

Feature does not have an \\"id\\" member.

Received:  [31m{\\"geometry\\": null, \\"properties\\": null, \\"type\\": \\"Feature\\"}[39m"
`;

exports[`Error Snapshot Testing. Throws error: Non-numeric search ID 1`] = `
"[2mexpect([22m[31mFeatureObject[39m[2m).toHaveNumericID([22m[32mSearchID[39m[2m)[22m

Did not find the numeric ID 1.

Received:  [31m{\\"geometry\\": {\\"coordinates\\": [[[100, 0], [101, 0], [101, 1], [100, 1], [100, 0]]], \\"type\\": \\"Polygon\\"}, \\"id\\": 1, \\"properties\\": null, \\"type\\": \\"Feature\\"}[39m"
`;

exports[`Error Snapshot Testing. Throws error: String ID 1`] = `
"[2mexpect([22m[31mFeatureObject[39m[2m).toHaveNumericID([22m[32mSearchID[39m[2m)[22m

ID is a string, expected a number.

Received:  [31m{\\"geometry\\": null, \\"id\\": \\"22\\", \\"properties\\": null, \\"type\\": \\"Feature\\"}[39m"
`;

exports[`Error Snapshot Testing. Throws error: Valid use case passes 1`] = `
"[2mexpect([22m[31mFeatureObject[39m[2m).not.toHaveNumericID([22m[32mSearchID[39m[2m)[22m

Expected input to not be a valid GeoJSON Feature object with number type ID.

Received:  [31m{\\"geometry\\": {\\"coordinates\\": [[[100, 0], [101, 0], [101, 1], [100, 1], [100, 0]]], \\"type\\": \\"Polygon\\"}, \\"id\\": 755, \\"properties\\": null, \\"type\\": \\"Feature\\"}[39m"
`;
